Add recipe for inform-mode

rrthomas opened the pull request:
### Brief summary of what the package does

inform-mode is an editing mode for Inform 6 files. (Inform 6 is a widely-used programming language for writing interactive fiction.)

### Direct link to the package repository

https://github.com/rrthomas/inform-mode

### Your association with the package

I’m the (new) maintainer (taking over from Rupert Lane).

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

One note: package-lint gives a warning:

    63:0: error: Customization groups should not end in "-mode" unless that name would conflict with their parent group.

I have not changed this, because inform-mode is a 30-year-old package, and it doesn’t seem worth breaking backwards compatibility at this point.

progfolio commented:
Very cool. Are you aware of https://github.com/speedenator/malyon? It'd be nice to see that package maintained as well.

rrthomas commented:
@progfolio I wasn't aware of it! Nice, but I don't play IF in Emacs, and I only used `inform-mode` nearly 20 years ago. I took over its maintenance because I want to continue to be able to edit my old code.

riscy commented:
Humongous thank you, @rrthomas, for this act of maintainership and for getting this up on MELPA.  Some small feedback -

⸺ `inform-mode.el` with package-lint 20250324.1912:
```
1 issue found:
63:0: error: Customization groups should not end in "-mode" unless that name would conflict with their parent group.
```
Before taking the plunge and indexing this on MELPA, I might make this (possibly breaking) change, just so that you and users can get ahead of it.

I also recommend (but won't insist) that you modernize the `defun inform-mode` and use `define-derived-mode` (deriving from `prog-mode`), which will handle a ton of the boilerplate you're carrying, e.g. `kill-all-local-variables`, `use-local-map`, `set-syntax-table`, *both* the creation of and automatic running of `inform-mode-hook`, etc.  It will lead to massive simplification and make it easier for the code to stay correct.

One other quick fix I see is in the definition of the keymap:
```
(defvar inform-mode-map nil
  "Keymap for Inform mode.")

(if inform-mode-map nil
...
```
which can be rewritten as:
```elisp
(defvar inform-mode-map 
  (let ((map (make-sparse-keymap)))
    (define-key inform-mode-map "\C-m" 'newline-and-indent)
    ;; ...
    )
  "Keymap for Inform mode.")
```
because `defvar` (unlike `setq`) will not change the value of the variable if it has already been set.

⸺ `inform-mode.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- inform-mode.el#L265: Prefer `defvar-local`, or `defcustom` with `:local t`
- inform-mode.el#L1614: Prefer `(add-to-list 'auto-mode-alist ...)`
- inform-mode.el#L1620: Loading a package should rarely add hooks
```
For the `rarely add hooks`, I would just delete this - users can decide themselves if they want font-lock on.

⸺ Package and license:
- In case you haven't, ensure GitHub release v1.6.2 is up-to-date with your current code and `Package-Version`

rrthomas commented:
> Humongous thank you, @rrthomas, for this act of maintainership and for getting this up on MELPA. Some small feedback -

Thanks as ever for your detailed feedback.

> ```
> 1 issue found:
> 63:0: error: Customization groups should not end in "-mode" unless that name would conflict with their parent group.
> ```
> 
> Before taking the plunge and indexing this on MELPA, I might make this (possibly breaking) change, just so that you and users can get ahead of it.

I addressed this in my initial comments. On reflection, I might as well get it over with. Since a user will have to install from MELPA to be affected by it, it's unlikely to take anyone greatly by surprise, and I think it's allowable to make this sort of breaking change once in 30 years.

> I also recommend (but won't insist) that you modernize the `defun inform-mode` and use `define-derived-mode` (deriving from `prog-mode`), which will handle a ton of the boilerplate you're carrying, e.g. `kill-all-local-variables`, `use-local-map`, `set-syntax-table`, _both_ the creation of and automatic running of `inform-mode-hook`, etc. It will lead to massive simplification and make it easier for the code to stay correct.

This is an excellent point. I would rather make this change *after* the first MELPA release. This follows my general process for taking over mature packages, which is first to make a clean-up release (what that means varies depending on my familiarity with the stack in question; for C I might do equivalent things, but in ELisp I'd rather not), and then make breaking and bigger changes. This hopefully makes it obvious to me which problems are pre-existing, and which are "my fault". The advantages are largely psychological, but it helps me to make an epoch *after* rather than *before* my first release.

> One other quick fix I see is in the definition of the keymap:
> 
> ```
> (defvar inform-mode-map nil
>   "Keymap for Inform mode.")
> 
> (if inform-mode-map nil
> ...
> ```
> 
> which can be rewritten as:
> 
> ```emacs-lisp
> (defvar inform-mode-map 
>   (let ((map (make-sparse-keymap)))
>     (define-key inform-mode-map "\C-m" 'newline-and-indent)
>     ;; ...
>     )
>   "Keymap for Inform mode.")
> ```
> 
> because `defvar` (unlike `setq`) will not change the value of the variable if it has already been set.

Thanks, I'll do that.

> ⸺ `inform-mode.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - inform-mode.el#L265: Prefer `defvar-local`, or `defcustom` with `:local t`
> - inform-mode.el#L1614: Prefer `(add-to-list 'auto-mode-alist ...)`
> - inform-mode.el#L1620: Loading a package should rarely add hooks
> ```

I'll address these.

>     * In case you haven't, ensure GitHub release v1.6.2 is up-to-date with your current code and `Package-Version`

Good point, I'll need to make a 1.6.3 release.

rrthomas commented:
I have gone all-in, and fixed all the issues you mentioned, removed some now-redundant functionality (for spelling and tags), and modernised various aspects of the code. I have therefore bumped the major version. Linting as per MELPA instructions is still clean.

riscy commented:
Thank you!

riscy commented:
And, in case you haven't, ensure GitHub release v2.0.0 is up-to-date with your current code and `Package-Version`

rrthomas commented:
@riscy many thanks. I have not made a release on GitHub, but I have tagged 2.0.0; a tag seems to suffice for other MELPA packages I maintain?

riscy commented:
Tags are equivalent to releases in GitHub's nomenclature. :)
